google/cloud/storage/retry.py

```diff
@@ -14,19 +14,21 @@
 
 import requests
 
-from google.api_core import exceptions
+from google.api_core import exceptions as exceptions_api
 from google.api_core import retry
+from google.auth import exceptions as exceptions_auth
 
 import json
 
 
 _RETRYABLE_TYPES = (
-    exceptions.TooManyRequests,  # 429
-    exceptions.InternalServerError,  # 500
-    exceptions.BadGateway,  # 502
-    exceptions.ServiceUnavailable,  # 503
-    exceptions.GatewayTimeout,  # 504
+    exceptions_api.TooManyRequests,  # 429
+    exceptions_api.InternalServerError,  # 500
+    exceptions_api.BadGateway,  # 502
+    exceptions_api.ServiceUnavailable,  # 503
+    exceptions_api.GatewayTimeout,  # 504
     requests.ConnectionError,
+    exceptions_auth.TransportError,
```

Comment:
According to https://github.com/googleapis/google-auth-library-python/blob/36e6f0feb41e901effc854a4f8c907deb9998f21/google/auth/exceptions.py#L22, `TransportError` is "Used to indicate an error occurred during an HTTP request". Looks like it can be thrown in a bunch of places. Is this too broad? Do we need to subclass this on the auth library side to narrow it down, or something?

Reply:
Doesn't look like there are any subclasses we can use.

https://github.com/googleapis/google-auth-library-python/blob/f1fee1f4c3d511d9e6ecbc1c0397e743bf2583db/google/auth/transport/requests.py#L187-L189

Maybe we could look at the inner exception, but `TransportError` does seem to be as specific as it gets.

Reply:
Yeah, I was wondering whether we should possibly add subclasses in the auth library to allow more differentiation. @frankyn could you speak more to what kind of problems will cause this exception to be raised?

Reply:
When I dug into this, `auth.TransportError` is only used within the auth package. However, as you mentioned it could include any of the following exceptions for at least requests: https://docs.python-requests.org/en/master/_modules/requests/exceptions/ but we only want to retry on ConnectionError.

Taking a note from the Go implementation, I'm sending another commit to instead unwrap the inner exception given we already retry on requests.ConnectionError.

Reply:
A bit strange that we are seeing requests.ConnectionError bare, but we're also seeing it wrapped by auth.

Reply:
We still have a feature request to wrap ConnectionError like ResetConnection in [api_core](https://github.com/googleapis/python-api-core/issues/60), however that doesn't resolve it for the auth library because it does its own wrapping.